File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        set<vector<int>>ans;
        vector<vector<int>>a1;
        sort(nums.begin(),nums.end());
        for(int i= 0 ;i < nums.size() ; i++){
            int j = i + 1 ;
            int k = n - 1 ;
            while(j<k){
                int su1 = nums[i] + nums[j] + nums[k];
                if(su1 == 0 ){
                    vector<int>temp;
                    temp.push_back(nums[i]);
                    temp.push_back(nums[j]);
                    temp.push_back(nums[k]);
                    ans.insert(temp);
                    // j +=1 ;
                    // k -=1 ;
                    while(j < n-1 and nums[j] == nums[j+1])        j += 1 ;
                    while(k > 0 and nums[k] == nums[k-1] )         k -= 1 ;
                    j += 1 ;
                    k -= 1 ;
                }
                else if(su1  < 0){
                    j += 1;
                }
                else{
                    k -= 1;
                }
            }
        }
        for(auto i : ans){
            a1.push_back(i);
        }
        return a1;
    }
};

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        nlist = nums.sort()
        print(nlist)
        i =0
        result_set = set()
        while i < n :
            j = i+1
            k = n-1
            while j<k:
                s1 = nums[i] + nums[j] + nums[k]
                if s1 == 0:
                    result_set.add((nums[i] , nums[j] ,nums[k]))
                    
                    while j<k and nums[j] == nums[j+1]:
                        j += 1
                    while j<k and nums[k] == nums[k-1]:
                        k -= 1

                    j += 1
                    k -=1
                elif s1 > 0 :
                    k-=1
                else :
                    j+=1
            i+=1

        return [i for i in list(result_set)]